jujugui/static/gui/src/app/store/env/go.js

```diff
@@ -555,36 +560,131 @@ YUI.add('juju-env-go', function(Y) {
+          return;
         }
-        this._send_rpc({
+        if (err) {
+          console.warn('macaroon authentication failed:', err);
+          return;
+        }
+        console.debug('macaroon authentication succeeded');
+      }.bind(this);
+
+      // Ensure this type of login is supported.
+      if (utils.compareSemver(this.get('jujuCoreVersion'), '2') === -1) {
+        cback('macaroon auth requires Juju 2');
+        return;
+      }
+
+      // Define the handler reacting to Juju controller login responses.
+      var handleResponse = function(bakery, macaroons, cback, data) {
+        if (data.Error) {
+          // Macaroon authentication failed or macaroons based authentication
+          // not supported by this controller. In the latter case, the
+          // controller was probably not bootstrapped with an identity manager,
+          // for instance by providing the following parameter to bootstrap:
+          // "--config identity-url=https://api.jujucharms.com/identity".
+          cback('authentication failed: ' + data.Error);
+          return;
+        }
+
+        var response = data.Response;
+        var macaroon = response['discharge-required'];
+        if (macaroon) {
+          // This is a discharge required response.
+          bakery.discharge(macaroon, (macaroons) => {
+            // Send the login request again including the discharge macaroon.
+            sendLoginRequest(
+              macaroons, handleResponse.bind(this, bakery, macaroons, cback));
+          }, (msg) => {
+            cback('macaroon discharge failed: ' + msg);
+          });
+          return;
+        }
+
+        // Macaroon authentication succeeded!
+        var user = response['user-info'] && response['user-info'].identity;
+        if (!user) {
+          // This is a beta version of Juju 2 which does not include user info
+          // in the macaroons based login response. Unfortunately, we did all
+          // of this for nothing.
+          cback('authentication failed: use a proper Juju 2 release');
+          return;
+        }
+        this.setCredentials({macaroons: macaroons, user: user});
+        cback(null, response);
+      };
+
+
+      // Define the function used to send the login request.
+      var sendLoginRequest = function(macaroons, callback) {
+        var request = {
           Type: 'Admin',
           Request: 'Login',
-          Params: params,
-          Version: version
-        }, this.handleLogin);
-        this.pendingLoginResponse = true;
-      } else {
-        console.warn('Attempted login without providing credentials.');
-        this.fire('login', {data: {result: false}});
-      }
+          Version: ADMIN_FACADE_VERSION_JUJU2
+        };
+        if (macaroons) {
+          request.Params = {macaroons: [macaroons]};
+        }
+        this._send_rpc(request, callback);
+      }.bind(this);
+
+      // Perform the API call.
+      var macaroons = this.getCredentials().macaroons;
```

Comment:
I _think_ it might be possible to get to this point with credentials being `undefined` and if that happens then this will cause the GUI to fall over. Maybe just add a conditional here just to be safe.